File: Sources/Runtime/Components/MelodySlider.swift
import SwiftUI
import Core

#if os(tvOS)
#else
/// Renders a continuous slider bound to a screen state key with configurable range.
struct MelodySlider: View {
    let definition: ComponentDefinition
    var onChanged: (() -> Void)? = nil

    @Environment(\.screenState) private var screenState

    var body: some View {
        Slider(
            value: binding,
            in: minValue...maxValue,
            step: stepValue
        )
        .melodyStyle(definition.style)
        .onChange(of: binding.wrappedValue) { _, _ in
            onChanged?()
        }
    }

    private var binding: Binding<Double> {
        guard let key = definition.stateKey else {
            return .constant(0)
        }
        return Binding(
            get: { screenState.get(key: key).numberValue ?? 0 },
            set: { screenState.set(key: key, value: .number($0)) }
        )
    }

    private var minValue: Double { definition.min ?? 0 }
    private var maxValue: Double { definition.max ?? 1 }
    private var stepValue: Double { definition.step ?? 0.1 }
}
#endif

File: Android/melody-runtime/src/main/java/com/melody/runtime/components/MelodySlider.kt
package com.melody.runtime.components

import androidx.compose.material3.Slider
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import com.melody.core.schema.ComponentDefinition
import com.melody.runtime.engine.LuaValue
import com.melody.runtime.renderer.LocalIsDisabled
import com.melody.runtime.renderer.LocalScreenState

@Composable
fun MelodySlider(definition: ComponentDefinition, onChanged: (() -> Unit)? = null) {
    val screenState = LocalScreenState.current
    val isDisabled = LocalIsDisabled.current
    val stateKey = definition.stateKey
    val minValue = (definition.min ?: 0.0).toFloat()
    val maxValue = (definition.max ?: 1.0).toFloat()
    val stepValue = definition.step?.toFloat() ?: 0f

    val currentValue = if (stateKey != null) {
        (screenState.slot(stateKey).value.numberValue ?: 0.0).toFloat()
    } else {
        0f
    }

    Slider(
        value = currentValue.coerceIn(minValue, maxValue),
        enabled = !isDisabled,
        onValueChange = { newValue ->
            if (stateKey != null) {
                screenState.set(stateKey, LuaValue.NumberVal(newValue.toDouble()))
            }
            onChanged?.invoke()
        },
        valueRange = minValue..maxValue,
        steps = if (stepValue > 0f) {
            ((maxValue - minValue) / stepValue).toInt() - 1
        } else {
            0
        },
        modifier = Modifier.melodyStyle(definition.style)
    )
}
